client/client_test.go

```diff
@@ -276,9 +274,117 @@ func TestNewEnvClientSetsDefaultVersion(t *testing.T) {
 	if client.version != expected {
 		t.Fatalf("Expected %s, got %s", expected, client.version)
 	}
+}
+
+// TestDowngradeVersionEmpty asserts that client.Client can
+// negotiate a compatible APIVersion when omitted
+func TestDowngradeVersionEmpty(t *testing.T) {
+	env := EnvToMap()
+	defer MapToEnv(env)
+
+	envMap := map[string]string{
+		"DOCKER_API_VERSION": "",
+	}
+	MapToEnv(envMap)
+
+	client, err := NewEnvClient()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ping := types.Ping{
+		APIVersion:   "",
+		OSType:       "linux",
+		Experimental: false,
+	}
+
+	// set our version to something new
+	client.UpdateClientVersion("1.25")
+
+	// if no version from server, expect the earliest
+	// version before APIVersion was implemented
+	expected := "1.24"
+
+	// test downgrade
+	client.DowngradeClientVersionPing(ping)
+	if client.version != expected {
```

Comment:
I was trying to use the same format as the rest of the tests in this suite. If I understood your suggestion, we want to use an assertion framework, so I reworked the tests here to use the same assertion library found in other packages here. I kept this in a separate commit, since it's unrelated cleanup.